### PR TITLE
feat(linter): typescript-eslint: prefer-function-type

### PR DESCRIPTION
**Pull request by zhangrunzhao**
Ref: #2180 

- doc: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-function-type.md
- code: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/prefer-function-type.ts
- test: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/prefer-function-type.test.ts

### CODE REVIEW

File: crates/oxc_linter/src/rules/typescript/prefer_function_type.rs [hunk 1]
```diff
@@ -0,0 +1,743 @@
...
+    /// // multiple call signatures (overloads) is allowed:
+    /// interface Overloaded {
+    ///   (data: string): number;
+    ///   (id: number): string;
+    /// }
+    /// // this is equivalent to Overloaded interface.
+    /// type Intersection = ((data: string) => number) & ((id: number) => string);
+    /// ```
+    PreferFunctionType,
+    correctness
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/prefer_function_type.rs:82]: According to https://typescript-eslint.io/rules/prefer-function-type, this is a `style` rule.

File: crates/oxc_linter/src/rules/typescript/prefer_function_type.rs [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, fixer::Fix, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("typescript-eslint(prefer-function-type): Enforce using function types instead of interfaces with call signatures.")]
+#[diagnostic(
+    severity(warning),
+    help("See https://typescript-eslint.io/rules/prefer-function-type for documentation.")
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/prefer_function_type.rs:18]: Can can improve the help message by saying what to do, e.g.

The function type form `() => string` is generally preferred when possible for being more succinct.

Comment (zhangrunzhao) [reply crates/oxc_linter/src/rules/typescript/prefer_function_type.rs:18]: image

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Thank you!

**Pull request merged**